### PR TITLE
Update crew API

### LINKED ISSUE #699: Update crew API
Should change current API routes and schemas listed here:
1. GET users/get-by-field
2. POST users/create
3. PUT users/edit-by-field
4. PUT users/edit-active

-> 

1. GET users/login
2. POST users/create - login required in data
3. PUT users/login
4. PUT users/login/edit-active

Also, make a supervisor optional.

### PR DESCRIPTION
**Pull request by DenisVorop**
## PR includes

-   [x] Feature

## Related issues

Resolve #699

### CODE REVIEW

File: src/trpc/router/restRouter.ts [hunk 1]
```diff
@@ -176,6 +177,297 @@ export const restRouter = router({
...
+                groupId: user.memberships.map((m) => m.groupId),
+                supervisorLogin: supervisor?.login,
+                active: user.active,
+            };
+        }),
+
+    editActiveByLogin: restProcedure
+        .meta({
+            openapi: {
+                method: 'GET',
```
Comment (9teen90nine) [inline src/trpc/router/restRouter.ts:429]: that should be a PUT

File: src/trpc/router/restRouter.ts [hunk 2]
```diff
@@ -16,9 +16,13 @@ import { getOrganizationUnitListSchema } from '../../modules/organizationUnitSch
 import { organizationUnitMethods } from '../../modules/organizationUnitMethods';
 import { getServiceListSchema } from '../../modules/serviceSchemas';
 import { serviceMethods } from '../../modules/serviceMethods';
+import { trimAndJoin } from '../../utils/trimAndJoin';
+import { config } from '../../config';
 
 import { tr } from './router.i18n';
 
+const getCorporateEmail = (login: string | null) => `${login}${config.corporateEmailDomain}`;
```
Comment (nyet-ty) [inline src/trpc/router/restRouter.ts:24]: what if there is no corporateEmailDomain in env?

File: src/trpc/router/restRouter.ts [hunk 3]
```diff
@@ -176,6 +206,265 @@ export const restRouter = router({
...
+                summary: 'Update user by login',
+            },
+        })
+        .input(
+            z.object({
+                login: z.string(),
+                data: z.object({
+                    surname: z.string(),
+                    firstName: z.string(),
+                    middleName: z.string().optional(),
```
Comment (nyet-ty) [inline src/trpc/router/restRouter.ts:224]: why only middle name optional in update?

**Review by nyet-ty — CHANGES_REQUESTED** [review]

File: src/trpc/router/restRouter.ts (hunk 3 above)
Comment (tsumo) [reply src/trpc/router/restRouter.ts:224]: Some people have only first name and surname, it's ok

File: src/trpc/router/restRouter.ts (hunk 2 above)
Comment (DenisVorop) [reply src/trpc/router/restRouter.ts:24]: will send only login

Comment (tsumo) [reply src/trpc/router/restRouter.ts:24]: loginundefined 🙂

**Review by 9teen90nine — APPROVED** [review]

File: src/trpc/router/restRouter.ts (hunk 3 above)
Comment (nyet-ty) [reply src/trpc/router/restRouter.ts:256]: I mean, why all other fields is mandatory? What if we want to update just the phone number?

Comment (tsumo) [reply src/trpc/router/restRouter.ts:256]: Agree, all fields should be optional then. Besides, we can't update just name or surname, it's stored in db as single string

File: src/trpc/router/restRouter.ts [hunk 4]
```diff
@@ -205,6 +238,215 @@ export const restRouter = router({
...
+                        data: {
+                            userId: user.id,
+                            serviceName: 'Phone',
+                            serviceId: input.data.phone,
+                        },
+                    });
+                }
+            }
+
+            return {
```
Comment (tsumo) [inline src/trpc/router/restRouter.ts:340]: Should create "edit user" history event before return

File: src/trpc/router/restRouter.ts [hunk 5]
```diff
@@ -205,6 +238,215 @@ export const restRouter = router({
...
+                corporateEmail: z.string(),
+                active: z.boolean(),
+                supervisorLogin: z.string().nullish(),
+            }),
+        )
+        .query(async ({ input }) => {
+            const user = await userMethods.getByLogin(input.login);
+            const updatedUser = await userMethods.editActiveState({ id: user.id, active: input.data.active });
+
+            return {
```
Comment (tsumo) [inline src/trpc/router/restRouter.ts:439]: "edit user active state" history event

**Review by tsumo — CHANGES_REQUESTED** [review]

File: src/trpc/router/restRouter.ts (hunk 4 above)
Comment (DenisVorop) [reply src/trpc/router/restRouter.ts:369]: fixed

File: src/trpc/router/restRouter.ts (hunk 5 above)
Comment (DenisVorop) [reply src/trpc/router/restRouter.ts:479]: fixed

File: src/trpc/router/restRouter.ts [hunk 6]
```diff
@@ -205,6 +238,255 @@ export const restRouter = router({
...
+            });
+
+            const phoneService = await prisma.userService.findFirst({
+                where: {
+                    userId: user.id,
+                    serviceName: 'Phone',
+                },
+            });
+
+            if (phoneService?.serviceName !== input.data.phone && input.data.phone) {
```
Comment (tsumo) [inline src/trpc/router/restRouter.ts:321]: phoneService?.serviceId !== input.data.phone

File: src/trpc/router/restRouter.ts [hunk 7]
```diff
@@ -205,6 +238,255 @@ export const restRouter = router({
...
+                            serviceId: input.data.phone,
+                        },
+                    });
+                }
+            }
+
+            const { before, after } = dropUnchangedValuesFromEvent(
+                {
+                    name: user.name,
+                    registrationEmail: user.email,
```
Comment (tsumo) [inline src/trpc/router/restRouter.ts:349]: Edit user event does not currently have all those properties

File: src/trpc/router/restRouter.ts [hunk 8]
```diff
@@ -205,6 +238,255 @@ export const restRouter = router({
...
+        )
+        .query(async ({ input, ctx }) => {
+            const user = await userMethods.getByLogin(input.login);
+            let newSupervisor: Awaited<ReturnType<typeof userMethods.getByLogin>> | null = null;
+
+            if (user.supervisor?.login !== input.data.supervisorLogin && input.data.supervisorLogin) {
+                newSupervisor = await userMethods.getByLogin(input.data.supervisorLogin);
+            }
+
+            const fieldsToUpdate: Record<string, string> = {};
```
Comment (tsumo) [inline src/trpc/router/restRouter.ts:283]: Let's create a new ticket for expanding userMethods.create. All this code can be moved to a method

**Review by tsumo — CHANGES_REQUESTED** [review]

File: src/trpc/router/restRouter.ts (hunk 8 above)
Comment (DenisVorop) [reply src/trpc/router/restRouter.ts:283]: https://github.com/taskany-inc/crew/issues/751

File: src/trpc/router/restRouter.ts (hunk 7 above)
Comment (DenisVorop) [reply src/trpc/router/restRouter.ts:349]: I added these fields to type for `editUser` event

https://github.com/taskany-inc/crew/pull/733/files#diff-2ecee2220e72fcc4a02d9a44eabad764811e295ca6cd0d86ab0f617a54125f99R21-R28

File: src/modules/historyEventTypes.ts [hunk 9]
```diff
@@ -18,7 +18,14 @@ interface HistoryEventsData {
         requireGroupId: false;
         requireUserId: true;
         requireBefore: true;
-        data: { name?: string; supervisorId?: string };
+        data: {
+            name?: string;
+            supervisorId?: string;
+            registrationEmail?: string;
```
Comment (tsumo) [inline src/modules/historyEventTypes.ts:24]: Just email

File: src/modules/historyEventTypes.ts [hunk 10]
```diff
@@ -18,7 +18,14 @@ interface HistoryEventsData {
...
+            phone?: string;
+            organizationalUnitId?: string;
+            supervisorLogin?: string;
```
Comment (tsumo) [inline src/modules/historyEventTypes.ts:27]: Probably unnecessary, since we already have id

File: src/modules/historyEventTypes.ts [hunk 11]
```diff
@@ -18,7 +18,14 @@ interface HistoryEventsData {
         requireGroupId: false;
         requireUserId: true;
         requireBefore: true;
-        data: { name?: string; supervisorId?: string };
+        data: {
```
Comment (tsumo) [inline src/modules/historyEventTypes.ts:21]: HistoryRecord component can be updated to display new values

**Review by tsumo — CHANGES_REQUESTED** [review]

**Review by tsumo — APPROVED** [review]

**Pull request merged**